File: milkshake.ts
/*
 * Reimplemented on July 9, 2022
 * Author: Gabriel Paiva
 */

class MailGenerator {
  SEP: Array<string> = ["", ".", "_"];

  nameList: Array<string> = [];
  domainList: Array<string> = [];

  combinedNames: Array<string> = [];
  result: Array<string> = [];

  splitName = (name: string) => {
    name = name.toLowerCase().trim();
    name = name.replace("/ (de|da|do|dos|das ) /", " ");
    let names = name.split(" ");
    return names;
  };

  addName = (name) => {
    this.nameList = this.splitName(name);
  };

  addDomain = (domains) => {
    domains.split(" ").map((domain) => {
      this.domainList.push(domain);
    });
  };

  run = () => {
    this.combineNames();
    this.combineDomains();
  };

  combineDomains = () => {
    for (let i in this.domainList) {
      for (let j in this.combinedNames) {
        // let filteredName = this.combinedNames[j].replace("_", "");
        // filteredName = filteredName.replace(".", "");
        let filteredName = this.combinedNames[j]
        if (filteredName.length <= 3) {
          continue;
        }
        this.result.push(`${filteredName}@${this.domainList[i]}`);
      }
    }
  };

  combineNames = () => {
    let names = this.nameList
    let separators = this.SEP
    let n = names.length
    let combinedNames = []
   
    // Caminhar nos nomes
    for (let i=0;i<n;i++){
        // Tamanho da janela definida pela quantidade de nomes
        for (let t=0;t<=n-i;t++){
            // Percorrer o final dos nomes
            for (let j=i;j<n;j++){ 
                if ((t)>n-j){
                    continue;  
                } 
                // Não repetir o nome sozinho (t=0)
                if (j==i && t>0){
                    continue;
                } 
                  for(let sepI in separators){
                    let separator = separators[sepI]
                      for (let w=0;w<=((t+1)*2)-1;w++){
                        let name
                          // Acumular os nomes	
                          if (w-(t+1)+1<=1 && 1<=w){	   		
                              name=names[i][0];
                          } else {
                              name=names[i];
                          }
                       let name2=name; // Reverte os nomes <sobrenome> <nome>	
                          for(let k=1;k<=t;k++){
                              if (w-(t+1)+1<=k+1 && k+1<=w){
                                  name+=separator+names[j+k-1][0];
                                  name2=names[j+k-1][0]+separator+name2;
                              } else {
                                  name+=separator+names[j+k-1];
                                  name2=names[j+k-1]+separator+name2;
                              }
                          }
                       combinedNames.push(name);
                       if ( name2 != name ){
                           combinedNames.push(name2);
                       }
                      }
                      if (t==0) break;		   		
                  }
                  if (t==0) break;
            }	
        }
    }
    this.combinedNames = combinedNames
    }
  }

const mb = new MailGenerator();
const names = "Sheldon Cooper";
const domains = "caltech.edu";
mb.addName(names);
mb.addDomain(domains);
mb.run()
console.log(mb.result)

File: milkshake.py
"""
Created on Sep 8, 2005 in PHP
Author: Ricardo Paiva

Reimplemented on July 9, 2022 in Typescript
Author: Gabriel Paiva

Rewritten on September 7, 2022 in Python
Author: Gabriel Paiva
"""

from sys import argv

class MailGenerator:
    def __init__(self):
        self.SEPARATORS = ['', '.', '_']
        self.CONNECTORS = ['du', 'de', 'des', "l'",
                           'la', 'le', 'da', 'do', 'dos', 'das', 'of']

        self.name_list = []
        self.domain_list = []

        self.combined_names = []
        self.combined_separators = set()
        self.result = []

    def add_name(self, full_name):
        filtered_full_name = full_name
        for connector in self.CONNECTORS:
            if connector.endswith("'"):
                # remove connectors ending with <<'>>
                filtered_full_name = filtered_full_name.replace(
                    f' {connector}', ' ')
            else:
                # remove others connectors
                filtered_full_name = filtered_full_name.replace(
                    f' {connector} ', ' ')

        # split full filtered name into array
        self.name_list = filtered_full_name.split(' ')

    def add_domains(self, domains):
        self.domain_list = domains.split(' ')

    def combine_names(self):
        # single names
        for name in self.name_list:
            self.combined_names.append([name])
        # full name
        self.combined_names.append(self.name_list)

        # remove one name
        for i in range(len(self.name_list)):
            half_name = self.name_list[:]
            half_name.pop(i)
            self.combined_names.append(half_name)

        for combined_name_i in range(len(self.combined_names)):
            for name_index in range(len(self.combined_names[combined_name_i])):
                # one single letter
                new_combination = self.combined_names[combined_name_i][:]
                new_combination[name_index] = new_combination[name_index][0]
                self.combined_names.append(new_combination)

                # one entire name
                new_combination = self.combined_names[combined_name_i][:]
                for name_index2 in range(len(new_combination)):
                    if name_index == name_index2:
                        continue
                    new_combination[name_index2] = new_combination[name_index2][0]
                self.combined_names.append(new_combination)

        # reversed
        for combined_name_i in range(len(self.combined_names)):
            reversed_combination = self.combined_names[combined_name_i][::-1]
            self.combined_names.append(reversed_combination)

    def combine_separators(self):
        for separator in self.SEPARATORS:
            for combination in self.combined_names:
                self.combined_separators.add(separator.join(combination))

    def combine_domains(self):
        for domain in self.domain_list:
            for combination in self.combined_separators:
                self.result.append(f'{combination}@{domain}')

    def generate(self, names, domains):
        self.add_name(names)
        self.add_domains(domains)
        self.combine_names()
        self.combine_separators()
        self.combine_domains()
        return self.result


